Improvements - Was "Communication with CentralReport Online"

Guimove opened the pull request:
## Update 03/03/2014

CentralReport Online is not planned anymore. Another project will be born in few days and will be announced on the [CR blog](http://blog.centralreport.net).
As this pull request adds useful features and many improvements in the core of this project, this branch has been merged into develop.

---
### Original pull request:

This pull request adds the communication with CentralReport Online (centralreport.net).
CentralReport Online is an online tool to follow all your hosts, in a single place. This part is currently in development and will be enhanced in the next weeks.
- [x] Add "requests" library
- [x] Create the "webservice" module
- [x] Create webservices entities
- [x] Clean and code optimization 

Remains to be done:
- [x] Clean values from the collector ('\n' in some variables, or too many spaces in others)
- [x] Clean useless entities ("Full", "Registration")
- [x] Compact the JSON output (on a single line, without useless spaces)
- [x] Cast dates and float to long (like disk spaces or memory data)
- [x] Update the thread "Checks" to send data to the Online part
- [x] Do more tests with the JSON generation
- [x] Add the "Agent name" and "Agent version" in the host json.

_Updated tasks (miniche, 02/04/2013)_
_Updated tasks (miniche, 29/05/2013)_
_Updated title and description (miniche, 30/07/2013)_

Guimove committed:
Insert of "requests" library and first play with web service

miniche committed:
Fixed requests zip

Guimove committed:
First full JSON object is done
Some optimization to do (read TODO comment)

miniche committed:
Removed the possibility of disabling some primary checks

miniche committed:
Cleaning web services

Guimove committed:
Check if the host is registered on centralreport.net

miniche committed:
Added new check entities

miniche committed:
Fixed indentation in "Check" entity

miniche committed:
Added support for HTTP verbs in web services

miniche committed:
Minor fixes in webservices functions

miniche committed:
Deleted useless code in web service thread

Must be done later in the feature-webservices branch.

miniche committed:
Updated class names

And some spelling fixes

miniche committed:
Fixed internal web server to work with new entities

Guimove committed:
Merge pull request #54 from CentralReport/feature-ws-entities

Working on new check entities

Guimove committed:
Creation of the serializer (convert an object in whatever you want)

Guimove committed:
Refactor serializer + remove serialize on objects

Guimove committed:
Typo mistake

miniche committed:
[WIP] Working on HATEOAS concept

miniche committed:
Working on web services.

Waiting for @Ninir to do the first test (JSON format for the HATEOAS).

miniche committed:
Removed useless "Webservices" class in the cr.webservice module

KISS! It's just a module, not a future object.

Guimove committed:
addition of the forgotten header

Ninir commented:
Please merge develop as it conflicts.

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-webservices

Conflicts:
	centralreport/cr/threads.py
	centralreport/web/server.py

Guimove committed:
Correction of the comments

Guimove committed:
Separation of serializer and object_converter classes

Guimove committed:
Serializer module rename

miniche committed:
Serializer module: merged "object_convert" into the serialize function

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-webservices

miniche committed:
Merge remote-tracking branch 'origin/feature-webservices' into feature-serializer

miniche committed:
Moved cr.webservices functions into cr.utils.web module.

New stuff this afternoon about web services and the online part, stay tuned!

miniche committed:
Merge pull request #57 from CentralReport/feature-serializer

Creation of the serializer (convert an object in whatever you want)

miniche committed:
Cleaned imports for the "webservices" module and the "requests" library

miniche committed:
Cleaned webservices methods for CentralReport Online

miniche committed:
Added disk UUID support

miniche committed:
Added: static files and 'data' module.

We have a new module, called 'data'. It contains data about the current host and the daemon. For now, just used to store the last check (cr.entities.checks.Check) and the host_info object (cr.entities.host.Info).
The static files can contains… static files, used for the web services, like the JSON templates for the exchanges with the Online part.

miniche committed:
Added 'errors' module and improved 'online' module

The errors module contains all definitions about the CentralReport custom errors. They can be used with the 'raise' function.
The online module have been cleaned and improved.

miniche committed:
Added 'try/except' statement when CentralReport executes a new check, avoiding daemon crashes.

miniche committed:
Cleaned the import of third-party libraries.

Now, the libraries are added into the sys.path in the centralreport module.

miniche committed:
Added docstrings to some variables for their types

This is useful for an IDE and for a developer to check the variable type quickly.

miniche committed:
Added attributes mapping for the Online web services.

miniche commented:
Remains to be done:
- Clean values from the collector ('\n' in some variables, or too many spaces in others)
- Clean useless entities ("Full", "Registration" and "Disks")
- Compact the JSON output (on a single line, without useless spaces)
- Cast dates and float to long (like disk spaces or memory data)
- Update the thread "Checks" to send data to the Online part
- Do more tests with the JSON generation
- Add the "Agent name" and "Agent version" in the host json.

Ninir commented on centralreport/cr/collectors.py:
```diff
@@ -419,29 +425,54 @@ def get_disks(self):
-                check_disk.used = disk_used
-                check_disk.free = disk_free
+                check_disk.name = disk_name.replace('/dev/', '')
+                check_disk.unix_name = disk_name
```
I'm not sure about `unix_name` here. As we have checks for UNIX only this does not matter. However, we'll probably have Windows Checks soon... Is there a specific reason?

miniche replied:
No specific reason, except that the Python version will be only designed for Unix-like systems.
On Windows, this name doesn't exist and replaced by a drive letter (like 'C:' or 'Z:', etc.).
If you want to standardize this, no problem, but we have to found a generic name for this attribute.

Ninir replied:
disk_name seems good :P

miniche replied:
In this case, we must rename the attribute "name" in the same class.
We could rename "name" to "display_name" and "unix_name" to "name".

Example, for OS X:

```
name = 'disk0s2'
display_name = 'Macintosh SSD'
```

(currently, it's `name = 'Macintosh SSD'` and `unix_name = 'disk0s2'`)

Ninir replied:
displayName seems fine to me too :)

miniche replied:
Great :)

Ninir commented on centralreport/cr/entities/webservices.py:
```diff
@@ -0,0 +1,44 @@
+    CentralReport - Webservices module
+        Contains all entities used with webservices.
+
+    https://github.com/miniche/CentralReport/
```
Mmmm... let's say no :D

Ninir commented on centralreport/cr/errors.py:
```diff
@@ -0,0 +1,30 @@
+
+"""
+    CentralReport - Errors module
+        Contains all definitions of the CentralReport custom errors
```
If errors belongs to CentralReport, then they are implicitly custom. This, I would remove the "custom" word

Ninir commented on centralreport/cr/online.py:
```diff
@@ -0,0 +1,277 @@
+
+"""
+    CentralReport - Online module
+        Contains all classes and functions used for the interaction with the Online server
```
`Contains all classes and functions interacting with the Online server`

Ninir commented on centralreport/cr/online.py:
```diff
@@ -0,0 +1,277 @@
+    json_vars['hostname'] = data.host_info.hostname
+    json_vars['displayName'] = data.host_info.hostname
+    json_vars['model'] = data.host_info.model
+    json_vars['cpuCount'] = data.host_info.cpu_count
```
What about the cpuModel as in the last file you added on DropBox?

miniche replied:
It's an oversight. I forgot to add this item in the documentation. (http://docs.centralreport.net/cro/api/users/hosts#postapiuserskeyhosts). Will be fixed tonight.

Ninir replied:
Hope so otherwise you'll be thrown into abysses with piranhas :D

miniche replied:
:fearful:

Ninir commented on centralreport/cr/online.py:
```diff
@@ -0,0 +1,277 @@
+    json_vars['kernelVersion'] = data.host_info.kernel_version
+    json_vars['architecture'] = data.host_info.architecture
+    json_vars['agent'] = 'TODO'
+    json_vars['agentVersion'] = 'TODO'
```
You are not forced to add them.. :D

miniche replied:
It's just to avoid forgetting it :D

Ninir replied:
You are not forced to remember them too... :D

Ninir commented on centralreport/cr/online.py:
```diff
@@ -0,0 +1,277 @@
+    elif ws_checks.code == 409:
+        raise OnlineError(3, 'Check already sent!')
+    elif ws_checks.code != 200:
+        raise OnlineError(4, 'Unkown answer code %s!' % ws_checks.code)
```
Unknown*

Ninir replied:
Btw, the server will return a 201 code as this is a new ressource added

miniche replied:
Mmh, ok. We must update the documentation in this case (http://docs.centralreport.net/cro/api/users/hosts/checks).

Ninir commented on centralreport/cr/static/online/send_checks.json:
```diff
@@ -0,0 +1,38 @@
+            "resident": {{ memory_resident }},
+            "free": {{ memory_free }},
+            "swapSize": {{ swap_size }},
+            "swapUsed": {{ swap_used }},
```
usedSwap

Ninir commented on centralreport/cr/static/online/send_checks.json:
```diff
@@ -0,0 +1,38 @@
+            "free": {{ memory_free }},
+            "swapSize": {{ swap_size }},
+            "swapUsed": {{ swap_used }},
+            "swapFree": {{ swap_free }}
```
freeSwap

Ninir commented on centralreport/cr/threads.py:
```diff
@@ -59,49 +50,53 @@ def __init__(self):
 
-                # Updating last check date...
+        # Checking CPU
+        log.log_debug('Doing a CPU check...')
```
Performing is more adapted ;)

Ninir commented on centralreport/cr/static/online/send_checks.json:
```diff
@@ -0,0 +1,38 @@
+            "active": {{ memory_active }},
+            "inactive": {{ memory_inactive }},
+            "resident": {{ memory_resident }},
+            "free": {{ memory_free }},
```
What do you think of using `freeMemory` (as for the tmpl_var)

miniche replied:
Are you talking about the key (`"free"`) or about the variable name (`{{ memory_free }}`)?

Ninir replied:
The key :)

miniche committed:
Fixed typo and links

miniche committed:
Added 'Agent name' and updated 'register_host' web services for CR Online

miniche commented:
Updated pull request description.

miniche committed:
Removed unused webservices entities

miniche committed:
Cleaned strings in the collectors

miniche committed:
Improved cr.utils.text functions

miniche committed:
Improved online module (parsing data, fixed wrong variables)

miniche committed:
Updated "Disk" entity (new 'name' and 'display_name' attributes)

Updated web service JSON to work with the new "Disk" entity attributes and with the new format (http://docs.centralreport.net/cro/api/users/hosts/checks)
Updated internal web server and collectors with the 'name' and 'display_name' attributes for the Disk entity

miniche committed:
Updated date format in the web services (CentralReport Online)

http://docs.centralreport.net/cro/api/users/hosts/checks

miniche committed:
Fixed docstring and added comments in the online module

miniche committed:
Removed useless keys in the configuration file (for the Online part)

miniche committed:
Added registration for the "disk resource" (CentralReport Online API)

miniche committed:
Added support for the new route (disk check), according to the API documentation.

http://docs.centralreport.net/cro/api/users/hosts/disks/checks

miniche committed:
Fixed attributes for web services with the Online server.

Guimove commented:
Can you merge develop in web services ?

miniche commented:
Sure! Will be done this evening.

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-webservices

Conflicts:
	centralreport/centralreport.py
	centralreport/cr/collectors.py

miniche committed:
Updated online web services: disk checks come back into the main check WS

miniche commented:
The official documentation (http://docs.centralreport.net/cro/api) has been updated according to c11f02c

miniche committed:
Removed "display_name" attribute

miniche committed:
Minor fixes in web services functions

miniche committed:
Fixed Webservices after tests

miniche committed:
Updated NEWS file for the pull request #52 (Communication with CentralReport Online)

miniche commented:
Hey guys! Great news... I think it's time to review this pull request :D

Ninir commented on NEWS:
```diff
@@ -8,6 +8,10 @@
     Currently in development
 
+    - Feature #52: Communication with CentralReport Online
+        * CentralReport Online is an online tool to follow all your hosts, in
```
`Follow` has a different meaning. What do you think of using `to check all your hosts` or `to get information of all your hosts`?

miniche replied:
True.. I will fix it tonight ;-)

Ninir commented on centralreport/cr/online.py:
```diff
@@ -0,0 +1,307 @@
+# They are empty when CR is starting, they will be gotten dynamically from the server.
+# TODO: Use HATEOAS for "route_host_check" when the CentralReport Online API will be updated
+route_user_check = 'http://centralreport.net/api/users/%key%'
+route_host_check = 'http://centralreport.net/api/users/%key%/hosts/%uuid%'
```
Why are you using this? Are you waiting for the `host_search` HATEOAS route?

miniche replied:
Yes! I have added this route to perform first tests. I have added a TODO (line 25) too, to remove this URL when the `host_search` will be available.

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-webservices

Conflicts:
	centralreport/centralreport.py
	centralreport/cr/collectors.py
	centralreport/cr/entities/host.py
	centralreport/cr/threads.py
	centralreport/cr/tools.py
	centralreport/web/server.py

miniche committed:
Minor fixes after merging develop into feature-webservices

Ninir commented:
Build failed :(

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-webservices

Conflicts:
	NEWS

miniche committed:
Added libraries autoloader

Third party libraries are now loaded when CentralReport is starting

miniche committed:
Fixed libraries loader in tests

miniche commented:
@Ninir Fixed since fed995d :)

miniche committed:
Added "config" argument to modify the configuration in the CLI.

miniche committed:
Removed Wizard during installation process

miniche commented:
This pull request will be cleaned in order to be merged in the develop branch, because this branch fixes and adds important stuff.
All related code to CentralReport Online will be removed, as this project will be reborn.

You can follow [@Central_Report](https://twitter.com/Central_Report/) on Twitter, we will tweet details about the future :)

miniche committed:
Cleaned Online references and code
